src/Microsoft.Identity.Client/Internal/Cache/RefreshTokenCacheItem.cs

```diff
@@ -33,21 +34,80 @@ namespace Microsoft.Identity.Client.Internal.Cache
     [DataContract]
     internal class RefreshTokenCacheItem : BaseTokenCacheItem
     {
+
         public RefreshTokenCacheItem()
         {
         }
 
-        public RefreshTokenCacheItem(string authority, string clientId, TokenResponse response) : base(authority, clientId, response)
+        public RefreshTokenCacheItem(string environment, string clientId, TokenResponse response) : base(clientId)
         {
             RefreshToken = response.RefreshToken;
+            Environment = environment;
+            PopulateIdentifiers(response);
+
+            User = User.Create(DisplayableId, Name, IdentityProvider,
+                GetUserIdentifier());
         }
 
+        [DataMember(Name = "environment")]
+        public string Environment { get; set; }
+
+        [DataMember(Name = "uid")]
+        public string Uid { get; set; }
+
+        [DataMember(Name = "utid")]
+        public string Utid { get; set; }
```

Comment:
Yep, this is exactly the behavior that will happen. It is a an edge case that should never happen unless there is a security issue and MSA needs to roll their salt/algorithm. This only applies to MSA btw.